# HarvestStat Data Profiling - Liberia

In [1]:
import glob, json
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
import requests
import numpy as np
import pandas as pd
import geopandas as gpd
from tools import save_npz, PrintAdminUnits, PlotAdminShapes
from tools import FDW_PD_Sweeper, FDW_PD_AvalTable, FDW_PD_Compiling, FDW_PD_ValidateFnidName
from tools import FDW_PD_CreateAdminLink, FDW_PD_RatioAdminLink, FDW_PD_ConnectAdminLink
from tools import FDW_PD_CaliSeasonYear
from tools_graphic import PlotBarProduction, PlotLinePAY
warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)
pd.options.mode.chained_assignment = None

In [2]:
# Load FEWS NET administrative boundaries ------- #
epsg = 'epsg:32629' # Liberia
fn_shapes = sorted(glob.glob('../data/shapefile/fewsnet/LR_Admin?_????.shp'))
shape_all = []
for fn in fn_shapes:
    name = fn[-18:-4]
    exec('%s = gpd.read_file("%s").to_crs("%s")' % (name, fn, epsg))
    exec('%s["area"] = %s["geometry"].area/10**6' % (name, name))
    exec('shape_all.append(%s)' % (name))
shape_all = pd.concat(shape_all, axis=0).reset_index(drop=True)
PrintAdminUnits(shape_all)
# ----------------------------------------------- #

# FDW API host address -------------------------- #
host = 'https://fdw.fews.net'
auth = tuple(json.loads(open('token.json', "r").read()))
parameters = {
    'format': 'json',
    'country': 'Liberia',
    'product':['R011','R012','R013','R014','R015','R017','R018'],
    'survey_type': 'crop:best'
}
endpoint = '/api/cropproductionindicatorvalue/'
response = requests.get(host + endpoint, auth=auth, params=parameters, proxies={})
response.raise_for_status()
df = pd.DataFrame.from_records(response.json())
# ----------------------------------------------- #

- FEWS NET admin shapefiles ------------------- #
| year	 | Admin1   | # units   | Admin2   | # units   | Admin3   | # units   |
| 1985	 | LR1985A1 | 13	| nan	| 0	| nan	| 0	|
| 2000	 | LR2000A1 | 14	| nan	| 0	| nan	| 0	|
| 2001	 | LR2001A1 | 15	| nan	| 0	| nan	| 0	|
| 2008	 | LR2008A1 | 15	| LR2008A2	| 136	| nan	| 0	|
----------------------------------------------- #


In [3]:
# Manual Pre-processing before Sweeping --------- #
# 1. Default setting 
# a) None-type crop production system
df.loc[df['crop_production_system'].isna(), 'crop_production_system'] = 'none'
df.loc[df['crop_production_system'] == '', 'crop_production_system'] = 'none'
# b) None-type population group
df.loc[df['population_group'].isna(), 'population_group'] = 'none'
df.loc[df['population_group'] == '', 'population_group'] = 'none'
# Replace season name
df['season_name'] = df['season_name'].replace({'Main harvest': "Main"})
# ----------------------------------------------- #

# FDW Production Data Inspection ---------------- #
df, df_raw = FDW_PD_Sweeper(df, area_priority='Area Planted')
table_dict = FDW_PD_AvalTable(df, shape_all)
# ----------------------------------------------- #

# FEWS NET Shapefile comparison ----------------- #
shape_used = pd.concat([
    LR_Admin1_1985, LR_Admin1_2008
], axis=0)
PlotAdminShapes(shape_used, label=True)
# ----------------------------------------------- #

- Remove missing records ---------------------- #
Orignial data points: 1,104
Removed 566 "Missing Value" points
0/276 "Area Harvested" points are retained.
131/276 "Area Planted" points are retained.
276/276 "Quantity Produced" points are retained.
131/276 "Yield" points are retained.
Current data points: 538

- Minor changes are applied ------------------- #

- Basic information --------------------------- #
Data period: 1995 - 2015
2 grain types are found: Cassava, Rice (Paddy)
1 seasons are found: Main (08-01)
1 crop production system are found: none
Data sources include:
[1] Ministry of Agriculture, Liberia --- FEWS NET Agro Maps, Liberia
[2] Ministry of Agriculture, Liberia --- Official Agricultural Statistics, Liberia
Administrative-1 fnids: 28
Administrative-2 fnids: 0
0 reporting units are found: 

- Total production over time ------------------ #
season_name    Main
                   
Cassava       59.3%
Rice (Paddy)  40.7%

- Crop calendar ------------------------------- #


- Liberia crop seasonal calendar

![T](../figures/crop_calendar/seasonal-calendar-liberia.png)

![USDA](https://ipad.fas.usda.gov/countrysummary/images/LI/cropcalendar/wafrica_li_calendar.png)

- FDW data consists of `LR1985A1` and `LR2008A1`.

| Year | Admin-1 | # units  | Admin-2  | # units |
| :---: | :----:  | :----:   | :----:   | :---:  |
| 1985	 | **`LR1985A1`** | 13	| -	| -	|
| 2000	 | LR2000A1 | 14	| -	| -	|
| 2001	 | LR2001A1 | 15	| -	| -	|
| 2008	 | **`LR2008A1`** | 15	| LR2008A2	| 136	|

- Comparison between admin boundaries.

![T](../figures/LR_admin_shapes.png)

- In 2008 (reporting), 2 districts are divided and added.

| 1985-2008|2008-present|
| :---:|:---:|
|LR1985A146 (Lofa) | LR2008A121 (Lofa), LR2008A145 (Gbarpolu)|
|LR1985A147 (Grand Gedeh) | LR2008A115 (Grand Gedeh), LR2008A142 (River Gee)|

- **`LR2008A1`** is used to represent current admin-level 1 crop data.
- Liberia has a single crop seasons: Main harvest (08-01).
- Liberia has no crop production system.
- Liberia has no population group.

In [4]:
# Define the latest shapefile ------------------- #
latest_level = 1
shape_latest = LR_Admin1_2008.copy().to_crs('epsg:4326')
# ----------------------------------------------- #

# Validation of FNIDs and Names ----------------- #
df = FDW_PD_ValidateFnidName(df, shape_used, shape_latest)
# ----------------------------------------------- #

# FDW Production Data Compiling ----------------- #
area, prod = FDW_PD_Compiling(df, shape_used)
area_all, prod_all = area.copy(), prod.copy()
mdx_pss = area.columns.droplevel([0,1]).unique()
# ----------------------------------------------- #

In [5]:
# Link admin boundaries ------------------------- #
link, over = FDW_PD_CreateAdminLink(LR_Admin1_1985, LR_Admin1_2008, 'ADMIN1', 'ADMIN1', prod, epsg)
assert all(np.unique([v['method'] for k,v in link.items()]) == 'PBR')
# Crop specific ratios
link_ratio = FDW_PD_RatioAdminLink(link, prod, over, mdx_pss)
# Add current unit to link_ratio
for fnid_new in link_ratio.keys():
    link_ratio[fnid_new][fnid_new] = 1.0
    link_ratio[fnid_new] = link_ratio[fnid_new].sort_index(axis=1, ascending=False)
# Connect data with AdminLink
area_new, prod_new = FDW_PD_ConnectAdminLink(link_ratio, area, prod, validation=True)
# ----------------------------------------------- #

In [6]:
# Manual correction ----------------------------- #
crop_new = prod_new/area_new
# ----------------------------------------------- #

# Complete long format DataFrame ---------------- #
df_area = area_new.T.stack().rename('value').reset_index()
df_area['indicator'] = 'area'
df_prod = prod_new.T.stack().rename('value').reset_index()
df_prod['indicator'] = 'production'
df_yield = crop_new.T.stack().rename('value').reset_index()
df_yield['indicator'] = 'yield'
stack = pd.concat([df_area, df_prod, df_yield], axis=0)
# Add "planting year"
cols = ['season_name','product','crop_production_system','planting_month','harvest_year','harvest_month','planting_year']
season_table = df[cols].drop_duplicates()
stack = stack.merge(season_table, on=cols[:-1])
# Add country and admin names
stack = stack.merge(df[['fnid','country','country_code','admin_1','admin_2']].drop_duplicates(), on='fnid', how='inner')
names = [
    'fnid','country','country_code','admin_1','admin_2','name',
    'product','season_name','planting_year','planting_month','harvest_year','harvest_month',
    'crop_production_system','indicator','value'
]
stack = stack[names]
# ----------------------------------------------- #

# Final Processing ------------------------------ #
# Manual change of product name
product_category_custom = {
    'Cassava':'Cassava',
    'Rice (Paddy)':'Rice',
}
stack['product'] = stack['product'].replace(product_category_custom)
for f, r in link_ratio.items():
    mdx = pd.MultiIndex.from_frame(r.index.to_frame().reset_index(drop=True).replace(product_category_custom))
    r.index = mdx
    link_ratio[f] = r
# Calibration of crop calendar using External Crop Calendar (ECC)
ecc = pd.read_csv('../data/crop_calendar/external_crop_calendar.csv')
ecc = ecc[ecc['country'] == 'Liberia']
stack, link_ratio = FDW_PD_CaliSeasonYear(stack, ecc, link_ratio)
# None admin names
stack.loc[stack['admin_1'].isna(), 'admin_1'] = 'none'
stack.loc[stack['admin_2'].isna(), 'admin_2'] = 'none'
# ----------------------------------------------- #

# Save data
fn_out = '../data/crop/adm_crop_production_LR.csv'
stack.to_csv(fn_out); print(f'{fn_out} is saved.')
save_npz('../data/crop/adm_crop_production_LR_ratio.npz', link_ratio)

All [season_name, product, crop_production_system] are in the external crop calendar.
../data/crop/adm_crop_production_LR.csv is saved.
../data/crop/adm_crop_production_LR_ratio.npz is saved.


# Visualization of production data

In [7]:
# Bar chart of national crop production
country_iso, country_name = 'LR', 'Liberia'
df = pd.read_csv('../data/crop/adm_crop_production_%s.csv' % country_iso, index_col=0)
df['year'] = df['harvest_year']
year = [df['year'].min(), df['year'].max()]
product_order = ['Cassava', 'Rice']
for season_name in ['Main']:
    footnote = 'National crop production in %s - %s' % (country_name, season_name)
    fn_save = '../figures/%s_bar_natgrainprod_%s.png' % (country_iso, season_name)
    sub = df[df['season_name'] == season_name]
    fig = PlotBarProduction(sub, year, product_order, footnote, fn_save)
    fig.show()

../figures/LR_bar_natgrainprod_Main.png is saved.


In [8]:
# Lineplot of Production-Area-Yield time-series
country_iso, country_name = 'LR', 'Liberia'
df = pd.read_csv('../data/crop/adm_crop_production_%s.csv' % country_iso, index_col=0)
df['year'] = df['harvest_year']
year = [df['year'].min(), df['year'].max()]
product_season = [
    ['Cassava','Main'],
    ['Rice','Main'],
]
for product_name, season_name in product_season:
    footnote = 'Production-Area-Yield time-series of %s - %s - %s' % (country_iso, product_name, season_name)
    fn_save = '../figures/%s_line_pay_%s_%s.png' % (country_iso, product_name, season_name)
    sub = df[(df['product'] == product_name) & (df['season_name'] == season_name)]
    fig = PlotLinePAY(sub, year, footnote, fn_save)
    fig.show()

../figures/LR_line_pay_Cassava_Main.png is saved.


../figures/LR_line_pay_Rice_Main.png is saved.
